crates/adapters/src/transport/http/output.rs

```diff
@@ -0,0 +1,227 @@
+
+/// `OutputTransport` implementation that sends data to websockets.
+pub struct HttpOutputTransport;
+
+impl OutputTransport for HttpOutputTransport {
+    fn name(&self) -> Cow<'static, str> {
+        Cow::Borrowed("http")
+    }
+
+    fn new_endpoint(
+        &self,
+        name: &str,
+        _config: &YamlValue,
+        async_error_callback: Box<dyn Fn(bool, AnyError) + Send + Sync>,
+    ) -> AnyResult<Box<dyn OutputEndpoint>> {
+        // let _config = HttpOutputConfig::deserialize(config)?;
+        let ep = HttpOutputEndpoint::new(name, async_error_callback)?;
+        Ok(Box::new(ep))
+    }
+}
+
+impl HttpOutputTransport {
+    pub(crate) fn get_endpoint_websocket(
+        endpoint_name: &str,
+        req: &HttpRequest,
+        stream: Payload,
+    ) -> AnyResult<HttpResponse> {
+        let endpoint = OUTPUT_HTTP_ENDPOINTS
+            .read()
+            .unwrap()
+            .get(endpoint_name)
+            .map(Clone::clone)
+            .ok_or_else(|| anyhow!("unknown HTTP output endpoint '{endpoint_name}'"))?;
+        if endpoint.num_sockets() >= MAX_SOCKETS_PER_ENDPOINT {
+            return Err(anyhow!(
+                "maximum number of connections per HTTP endpoint exceeded"
+            ));
+        }
+        let resp = ws::start(HttpOutputWs::new(endpoint), req, stream)
+            .map_err(|e| anyhow!(format!("error initializing websocket: {e}")))?;
+        info!("HTTP output endpoint '{endpoint_name}': opened websocket");
+        Ok(resp)
+    }
+}
+
+#[derive(Clone, Deserialize, ToSchema)]
+pub struct HttpOutputConfig;
+
+struct HttpOutputEndpointInner {
+    name: String,
+
+    /// Addresses of websocket actors associated with the endpoint.
+    ///
+    /// HTTP endpoint supports up to MAX_SOCKETS_PER_ENDPOINT simultaneously
+    /// open websockets.  The client opens a new websocket by issuing a GET
+    /// request to the `/output_endpoint/{endpoint_name}` endpoint.  The
+    /// websocket is removed from this set when client connection closes.
+    ///
+    /// This field is used to notify all websocket actors about new data
+    /// buffers to send out.
+    socket_addrs: RwLock<HashSet<Addr<HttpOutputWs>>>,
+    _async_error_callback: Box<dyn Fn(bool, AnyError) + Send + Sync>,
+}
+
+impl HttpOutputEndpointInner {
+    fn new(name: &str, async_error_callback: Box<dyn Fn(bool, AnyError) + Send + Sync>) -> Self {
+        Self {
+            name: name.to_string(),
+            socket_addrs: RwLock::new(HashSet::new()),
+            _async_error_callback: async_error_callback,
+        }
+    }
+}
+
+/// Output endpoint that establishes websocket connections with clients on
+/// demand and sends output batches to these websockets.
+///
+/// This implementation provides no support for reliable delivery
+/// and is mostly intended for browser-based testing.
+#[derive(Clone)]
+struct HttpOutputEndpoint {
+    inner: Arc<HttpOutputEndpointInner>,
+}
+
+impl HttpOutputEndpoint {
```

Comment:
could some of this be refactored into a base struct common to both Input and Output?